modules/core/src/main/java/org/apache/ignite/internal/processors/tracing/configuration/GridTracingConfiguration.java

```diff
@@ -0,0 +1,303 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.tracing.configuration;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.ignite.IgniteCheckedException;
+import org.apache.ignite.IgniteLogger;
+import org.apache.ignite.internal.GridKernalContext;
+import org.apache.ignite.internal.processors.metastorage.DistributedMetaStorage;
+import org.apache.ignite.internal.processors.metastorage.persistence.DistributedMetaStorageImpl;
+import org.apache.ignite.internal.processors.tracing.Scope;
+import org.apache.ignite.internal.util.tostring.GridToStringExclude;
+import org.apache.ignite.internal.util.typedef.internal.LT;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Tracing configuration implementation that uses distributed meta storage in order to store tracing configuration.
+ */
+public class GridTracingConfiguration implements TracingConfiguration {
+    /** */
+    private static final String TRACING_CONFIGURATION_DISTRIBUTED_METASTORE_KEY_PREFIX =
+        DistributedMetaStorageImpl.IGNITE_INTERNAL_KEY_PREFIX + "tr.config.";
+
+    /** Map with default configurations. */
+    private static final Map<TracingConfigurationCoordinates, TracingConfigurationParameters> DEFAULT_CONFIGURATION_MAP;
+
+    static {
+        Map<TracingConfigurationCoordinates, TracingConfigurationParameters> tmpDfltConfigurationMap = new HashMap<>();
+
+        tmpDfltConfigurationMap.put(
+            new TracingConfigurationCoordinates.Builder(Scope.TX).build(),
+            TracingConfiguration.DEFAULT_TX_CONFIGURATION);
+
+        tmpDfltConfigurationMap.put(
+            new TracingConfigurationCoordinates.Builder(Scope.COMMUNICATION).build(),
+            TracingConfiguration.DEFAULT_COMMUNICATION_CONFIGURATION);
+
+        tmpDfltConfigurationMap.put(
+            new TracingConfigurationCoordinates.Builder(Scope.EXCHANGE).build(),
+            TracingConfiguration.DEFAULT_EXCHANGE_CONFIGURATION);
+
+        tmpDfltConfigurationMap.put(
+            new TracingConfigurationCoordinates.Builder(Scope.DISCOVERY).build(),
+            TracingConfiguration.DEFAULT_DISCOVERY_CONFIGURATION);
+
+        DEFAULT_CONFIGURATION_MAP = Collections.unmodifiableMap(tmpDfltConfigurationMap);
+    }
+
+    /** Kernal context. */
+    @GridToStringExclude
+    protected final GridKernalContext ctx;
+
+    /** Grid logger. */
+    @GridToStringExclude
+    protected final IgniteLogger log;
+
+    /**
+     * Constructor.
+     *
+     * @param ctx Context.
+     */
+    public GridTracingConfiguration(@NotNull GridKernalContext ctx) {
+        this.ctx = ctx;
+
+        log = ctx.log(getClass());
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean addConfiguration(
```

Comment:
Personally, I'd name it "applyConfiguration"

Reply:
Agreed.